src/proposer.ts

```diff
@@ -121,6 +130,246 @@ let s: ReturnType<typeof strings>
+	}
+	if (intention.inputs.length !== intention.outputs.length) {
+		throw new Error(
+			'AssignDeposit requires 1:1 mapping between inputs and outputs'
+		)
+	}
+
+	// Zero-fee enforcement
+	if (!Array.isArray(intention.totalFee) || intention.totalFee.length === 0) {
+		throw new Error('AssignDeposit requires totalFee with zero amount')
+	}
+	const allTotalZero = intention.totalFee.every((f) => f.amount === '0')
+	if (!allTotalZero) {
+		throw new Error('AssignDeposit totalFee must be zero')
+	}
+	if (
+		Array.isArray(intention.proposerTip) &&
+		intention.proposerTip.length > 0
+	) {
+		throw new Error('AssignDeposit proposerTip must be empty')
+	}
+	if (
+		Array.isArray(intention.protocolFee) &&
+		intention.protocolFee.length > 0
+	) {
+		throw new Error('AssignDeposit protocolFee must be empty')
+	}
+	if (Array.isArray(intention.agentTip) && intention.agentTip.length > 0) {
+		throw new Error('AssignDeposit agentTip must be empty if provided')
+	}
+
+	for (let i = 0; i < intention.inputs.length; i++) {
+		const input: IntentionInput = intention.inputs[i]
+		const output: IntentionOutput = intention.outputs[i]
+
+		if (!output || (output.to === undefined && !output.to_external)) {
+			throw new Error('AssignDeposit requires outputs[].to (vault ID)')
+		}
+		if (output.to_external !== undefined) {
+			throw new Error('AssignDeposit does not support to_external')
+		}
+
+		if (input.asset.toLowerCase() !== output.asset.toLowerCase()) {
+			throw new Error('AssignDeposit input/output asset mismatch at index ' + i)
+		}
+		if (input.amount !== output.amount) {
+			throw new Error(
+				'AssignDeposit input/output amount mismatch at index ' + i
+			)
+		}
+		if (input.chain_id !== output.chain_id) {
+			throw new Error(
+				'AssignDeposit input/output chain_id mismatch at index ' + i
+			)
+		}
+
+		// Validate on-chain vault existence
+		await validateVaultIdOnChain(Number(output.to))
+	}
+}
+
+/**
+ * Discovers ERC-20 deposits made by `controller` into the VaultTracker and
+ * ingests them into the local `deposits` table. Uses Alchemy's decoded
+ * getAssetTransfers API for reliability and simplicity.
+ *
+ * If fromBlock/toBlock are not provided, computes a default lookback window
+ * of ~7 days by subtracting ~50,400 blocks from the latest block.
+ */
+async function discoverAndIngestErc20Deposits(params: {
+	controller: string
+	token: string
+	chainId: number
+	fromBlockHex?: string
+	toBlockHex?: string
+}): Promise<void> {
+	await discoverAndIngestDeposits({
+		controller: params.controller,
+		chainId: params.chainId,
+		categories: ['erc20'],
+		token: params.token,
+		fromBlockHex: params.fromBlockHex,
+		toBlockHex: params.toBlockHex,
+	})
+}
+
+/**
+ * Discovers ETH (internal) deposits made by `controller` into the VaultTracker
+ * and ingests them into the local `deposits` table via idempotent inserts.
+ */
+async function discoverAndIngestEthDeposits(params: {
+	controller: string
```

Comment:
See earlier note about leaving `controller` unspecified and getting all ETH deposits from any address.